src/NUnitFramework/framework/Internal/TestFilter.cs

```diff
@@ -172,16 +177,20 @@ public static TestFilter FromXml(TNode node)
             {
                 case "filter":
                 case "and":
-                    var andFilter = new AndFilter();
+                    List<TestFilter> childFilters = new List<TestFilter>();
+
                     foreach (var childNode in node.ChildNodes)
-                        andFilter.Add(FromXml(childNode));
-                    return andFilter;
+                        childFilters.Add(FromXml(childNode));
+
+                    return new AndFilter(childFilters.ToArray());
 
                 case "or":
-                    var orFilter = new OrFilter();
+                    List<TestFilter> orChildFilters = new List<TestFilter>();
+
                     foreach (var childNode in node.ChildNodes)
-                        orFilter.Add(FromXml(childNode));
-                    return orFilter;
+                        orChildFilters.Add(FromXml(childNode));
+
+                    return new OrFilter(orChildFilters.ToArray());
```

Comment:
Would it make sense to add an internal `IList<TestFilter>` overload on the constructors of `AndFilter`, `OrFilter`, and `CompositeFilter`? I am thinking it may save on the need to call `ToArray()` here + above since the `List<TestFilter>` can then be passed right through into the `ReadOnlyCollection<TestFilter>()` constructor

Reply:
I'm reluctant to do more API changes that slightly required. The ToArray call while unfortunate is comparably cheap to other things that happen during the test execution.

Reply:
Sure, that makes sense. Thanks for considering the feedback and for your response @pakrym .
I'm excited to see this change land, it's a huge improvement.